download_pod_logs/model.py

```diff
@@ -216,15 +216,36 @@ def _write_log(kubectl: KubectlInterface, pod: KubernetesResource, tail: int, ou
                 output_file.writelines(("#" * 50) + "\n\n")
 
                 # call kubectl to get the log for this container
-                try:
-                    log: List[AnyStr] = kubectl.logs(pod_name=f"{pod.get_name()} {container_status.get_name()}",
-                                                     all_containers=False, prefix=False, tail=tail)
 
+                log: List[AnyStr] = list()
+                try:
+                    log = kubectl.logs(pod_name=pod.get_name(), container_name=container_status.get_name(),
+                                       all_containers=False, prefix=False, tail=tail, ignore_errors=False)
                 except CalledProcessError:
-                    err_msg = (f"ERROR: A log could not be retrieved for the container [{container_status.get_name()}] "
-                               f"in pod [{pod.get_name()}] in namespace [{kubectl.get_namespace()}]")
-                    log: List[AnyStr] = [err_msg]
-                    err_info.append((container_status.get_name(), pod.get_name()))
+                    # container log has error, we'll retrieve log from initContainers
+                    try:
+                        log = kubectl.logs(pod_name=pod.get_name(),
+                                           container_name=container_status.get_name(),
+                                           all_containers=False, prefix=False, tail=tail,
+                                           ignore_errors=True)
+
+                        initcontainers: Optional[List[Dict]]
+                        initcontainers = pod.get_spec_value(KubernetesResource.Keys.INIT_CONTAINERS)
+
+                        if initcontainers:
+                            for initcontainer in initcontainers:
+                                name: Text = initcontainer.get("name")
+                                log = log + ["\n" + "#" * 50] + [f"# Log from initContainer: {name}"] + ["#" * 50]
+                                log = log + kubectl.logs(pod_name=pod.get_name(), container_name=name,
+                                                         all_containers=False, prefix=False, tail=tail,
+                                                         ignore_errors=True)
+
+                    except CalledProcessError:
+                        err_msg = (f"ERROR: A log could not be retrieved for the container "
+                                   f"[{container_status.get_name()}] "
+                                   f"in pod [{pod.get_name()}] in namespace [{kubectl.get_namespace()}]")
+                        log: List[AnyStr] = [err_msg]
+                        err_info.append((container_status.get_name(), pod.get_name()))
```

Comment:
I have a few recommended changes for this section, so I'm going to add an updated code snippet below to hopefully help alleviate any confusion. I'll discuss the reason for the changes below the snippet.

```python

try:
    log = kubectl.logs(pod_name=pod.get_name(), container_name=container_status.get_name(),
                                   all_containers=False, prefix=False, tail=tail, ignore_errors=False)
except CalledProcessError as e:
    # print to stdout that an error has occurred
    err_info.append((container_status.get_name(), pod.get_name()))
    
    # container log has error, we'll retrieve log from initContainers
    # add error from container call
    log = e.output

    initcontainers: Optional[List[Dict]] = pod.get_spec_value(KubernetesResource.Keys.INIT_CONTAINERS)

    if initcontainers:
        for initcontainer in initcontainers:
            name: Text = initcontainer.get("name")
            log = log + ["\n" + "#" * 50] + [f"# Log from initContainer: {name}"] + ["#" * 50]
     
            try:
                log = log + kubectl.logs(pod_name=pod.get_name(), container_name=name,
                                                         all_containers=False, prefix=False, tail=tail,
                                                         ignore_errors=False)
            except CalledProccessError as e:
                err_msg = (f"ERROR: A log could not be retrieved for the initcontainer "
                                    f"[{name}] "
                                    f"in pod [{pod.get_name()}] in namespace [{kubectl.get_namespace()}]")
                err_info.append((container_status.get_name(), pod.get_name()))

                log = log + e.output
```

A reference to the CalledProcessError caught on line 224 was added so that the second call to `kubectl.logs()` with `ignore_errors=True` is no longer needed. Setting `log = e.output` accomplishes the same thing as making that second call.

I added back in the append to `err_info`. Without this, there is nothing surfaced up the user that an issue occurred with the container, so they don't know to go look at the logs without it.

The nested try/except was moved to surround just the `kubectl.logs()` call for the initcontainer. Also I set `ignore_errors=False`. Without these two things, if an error occurred with the initcontainer, there would be no way of knowing. By adding in the error message for the container above, the user knows now to look at that particular log, this just provides more info in the event that an error occurs with the initcontainer as well. To that end, I reworked the `except` block to add an error message for the initcontainer too.

Reply:
code updated.